### ISSUE TITLE
Race Conditions for Requests

### ISSUE THREAD

**Issue by BenjaminHohlmann** [IHE/DEV.SDPi#359]
**Problem:**
SDC Messages are subject to race condition and their order is not guaranteed. If implemented without proper safeguards, this may lead to hazardous situations with unexpected behavior. 

There is no guarantee, that messages are transmitted sequentially over the network. 
Simultaneously, there is no requirement, that SDC participants wait for a request to finish before starting the next request.

**Exemplary error-case scenario** 
A standard-compliant consumer may issues a series of set operations on a metric without waiting for a confirmation before issuing the next set operation. Intuitively, the consumer may assume that the metric will be set to the value of the last set operation. However, due to the lack of guaranteed message ordering, this assumption might not hold true.

Notably, when using the regular HTTP request/response pattern over a single socket (as most implementations presumably do), this issue does not arise. Requests are buffered until a response is received, and only then can the next request be processed. However, problems may occur when communication is established over multiple sockets, for example, to improve performance. Similarly, this issue is relevant when using gRPC instead of MDPWS.

Additionally, this error case cannot be detected from the OperationInvoked reports. These reports provide information about the metric, operation handles, and invocation state, but not the value being set.

Likewise, setXXXResponse messages do not carry sufficient information to determine execution order, as the reported mdibVersion represents the version _before_ applying the set request.

**Solutions:**
**1) Add a requirement, stating that the request/response pattern is to be followed at all times.**
Most manufacturers likely already implement this pattern. It is a simple, safe, and backward-compatible solution that does not require significant changes.

On the other hand, this introduces limitations performance-wise: 
The queries-per-seconds (Qps) are limited to 1 / round-trip-time, with the latter being at least twice the latency. With a  latency of 100ms, this limits the queries-per-second to 5. Depending on the application, this could be a critical safety issue.
In general, it doubles the distributed system's vulnerability to latency compared to solution 2).

**2) Add a message-counter for each connection for the recipient to order the incoming messages.**
This requires introducing additional fields. 

**3) Adapt the OperationInvoked report, such that out-of-order executions can be detected.**
By reporting the _new_ mdibVersion, too, the order of execution could be determined.
Alternatively, the actual value could be included in the report.
This solves the issue just for set requests, not in general. It may still arise with other services, even though those already existent are not affected by it (needs confirmation).

**4) Add an _optional_ requirement, to make manufacturers aware of this issue, but maintain maximal flexibility in implementation**
As mentioned in 1), not waiting for a response improves overall system responsiveness. As such, it may be desired behavior. 
Manufacturers are hinted at this issue/opportunity by the requirement, reducing its risk.
Open question: If some manufacturers adhere to the requirement, and some don't, is the overall system still safe? Does this needs to be communicated during handshake?

**Comment by PaulMartinsen** [comment]
There are a couple of requirements in SDPi that could be relevant to this: 
* [R1003](https://profiles.ihe.net/DEV/SDPi/index.html#:~:text=R1003,-A%20SOMDS%20Participant) mandates HTTP 1.1 without pipelining. I think this means participants must wait for an HTTP response to every message before sending another message though, as you noted, this limits performance. 
* [R1001](https://profiles.ihe.net/DEV/SDPi/index.html#:~:text=present%20no%20gap.-,R1001,-A%20SOMDS%20Provider) mandates providers establish only a single TCP connection for each consumer. I think the intent here is to prevent providers sending parallel notifications over multiple connections, which could lead to the race conditions you mention. 

Do these requirements address the race you condition you raised? 

Some consumers are using the mdib version to check message ordering ([R1005](https://profiles.ihe.net/DEV/SDPi/index.html#vol1_spec_sdpi_p_actor_somds_consumer:~:text=R1005,-A%20SOMDS%20Consumer)), though I think there are [some problems with that](https://github.com/IHE/DEV.SDPi/issues/321). The solution proposed in that issue doesn't anticipate multiple connections though. Nonetheless, could it help resolve the race condition you raised while allowing improved performance?

**Comment by BenjaminHohlmann** [comment]
Thank you very much for pointing these out!

We think R1001 is meant to be applied to reports only, as indicated by its leading text, the Chapter and the fact that it can not be fulfilled for services, where the provider serves as host: 
Reports are send out by the provider's HTTP client and there can only be one TCP connection. Thus, any service that requires a HTTP server on the provider's side (including set!) would require opening a second TCP connection, violating the requirement.

R1003 is not sufficient, as one may still set up multiple HTTP 1.1 connections.

**Comment by BenjaminHohlmann** [comment]
The proposed solution of introducing a mechanism for "transmitted reports counting" for partial subscriptions does not solve the issue, as it applies to reports only, for which the issue does not apply, given R1001.
Yet, if the mechanism was extended to all messages, it would indeed implement solution 2).

**Comment by PaulMartinsen** [comment]
There's tricky language in all this, but I think you are right about R1001 being for out-bound connections only. 
> A [SOMDS Provider](https://profiles.ihe.net/DEV/SDPi/index.html#vol1_spec_sdpi_p_actor_somds_provider) shall only establish one TCP connection at a time for every subscribed [SOMDS Consumer](https://profiles.ihe.net/DEV/SDPi/index.html#vol1_spec_sdpi_p_actor_somds_consumer).

Strictly, once a connection's been established for notifications I think this requirement, as written now, says that one connection should be used for both sending notifications from to a consumer and receiving operation commands from that consumer. I wonder if that's the intent? I suspect it isn't, but if so would create a bunch of new questions, e.g.:
* should the connection stay open when the subscription ends in case there are more operation command messages?
* who's supposed to close the connection?

I agree that R1003 isn't sufficient to avoid race conditions if the consumer is permitted to setup multiple connections to a single provider; and I can't see anything that bans that. However, this may be the same race condition that happens when multiple consumers each setup a single connection to the provider and send conflicting commands. I'm only just starting to look at invoked operations now, so I'm not sure how that's resolved. There are a bunch of requirements in 10207, §7.4.2 that may be relevant though.

**Comment by PaulMartinsen** [comment]
Hi @BenjaminHohlmann, 

I do like your drill example. Makes me think of a dentist who can't manage to shut off his biceps drill. 

A couple of related questions I've been trying to figure out:
1. 11073-10207:§7.4.2 R0057 and R0056, as you noted, have consumers shall subscribe to `OperationInvokedReport` and providers should reject service operations if there isn't a subscription. However, there was a bit of discussion at the last SDPi workshop about how consumers would be identified for transport security; as far as I can tell consumers have no immutable identity. So I don't see how a provider would make sure a specific consumer has subscribed to the `OperationInvokedReport`. IP addresses seems insufficient if the consumer's machine is running multiple consumers. 
2. In the `OperationInvokedReport`, there is an `InvocationSource` element for participants to track who requested an operation. It also seems important for auditing. But again, it isn't clear where this information comes from. 

You noted transaction ids are "monotonically increasing integer assigned by the provider and included in the SetResponse". Just wondering where that came from? I thought they need to be unique across all transactions, at least within a `MdibVersionGroup/@SequenceId`. But I haven't seen anything describing the relationship between transactions with different ids. 

On the concurrency side, I guess there could be many consumers. So in your example, one consumer might instruct the drill to start and another for it to stop. I see R0058 and R0059 talking about concurrent consumers but I'm not sure it covers the case of one consumer behaving concurrently. Does this mean the consumer should wait until it gets a `InvocationState` = `Fin` after it sends the start command before sending the stop command? This is probably a minor quibble though. What if your surgeon accidentally starts the drill and a nurse wants to intervene to stop it with a separate foot switch?

20701, §7.2.1 introduces safety information but I'm still figuring this out myself. It might be possible to link the mdib version to safety information, to support your 3rd solution. That is, I think it may be possible to setup an operation that would only start the drill if the mdib version matched the version supplied by the consumer when the operation was received by the provider. 

Using timestamps to synchronize operations may be problematic. Two participants with clocks properly synchronized using, for example, the NTPv4 protocol could (theoretically) differ by more than [8 minutes](https://github.com/IHE/DEV.SDPi/issues/295#:~:text=be%20less%20than-,8%20minutes%20and%2032%20seconds,-.). On top of that, there could be a non-slewing time adjustment (indeterminant step change) might occur at a particularly inconvenient time. 

Also, do I understand correctly that the message counter you're suggesting for the drill example is managed by the consumer? Will that work if there are multiple concurrent consumers, each with their own message counters (and no identity)? 

Concurrency may be just the tip of the iceberg though. Even with the drill start and stop operations not happening out of order, communications between consumer and provider could fail at any time. Perhaps the provider has to ensure biceps operations are not the only way to shut off the drill?

**Comment by BenjaminHohlmann** [comment]
Yes, this may be just the iceberg's tip, yet it is an unpleasantly complex topic, already.

On the TransactionID, i was in fact not able to find any statement that requires it to be monotonically increasing. That's just the way we implemented it. As it is an unsigned integer, it is _the_ obvious, simple, fast and reliable way to do this. Yet, this should be made official.
Furthermore, the TransactionID is in no way linked to the actual execution. A provider may set the TransactionID for request A, then switch threads to another request B, set this one's TransactionID and execute it first, before returning to request A. The TransactionID needs to be linked to the operation invoked state machine, otherwise it is not useful for concurrency detection.

Re-visiting my proposed solutions, I would like to back up your reasoning, that there is only limited **synchronization**. The 8 minutes you mentioned are the maximum time discrepancy between two participants that will be corrected by slewing adjustments, correct?. Which does not mean, that we should expect up to 8 min of time discrepancy in real world scenarios. Rather the 'few milliseconds' you cite a bit down in your post. Still, time synchronization is not reliable and it's a huge source of errors and risks. So scrap that.
A consumer-provider-pair-wise **message counter** can only address the issue of single-consumer-multiple-sockest concurrency. 
The consumer **communicating the mdib version** it wants its change applied to is the only reliable option to detect concurrency in a multiple-concurrent-consumer scenario out of the three proposed solutions.

Reading through GLUE $7.2.1, the "safety information" you mention is exactly what we are looking for. According to [this paper](https://www.amd.e-technik.uni-rostock.de/veroeff/CSCN_2015_09_03_v01.pdf), the provider defines what information is required for his operations to be executed safely. An example of an OR table remotely controlled by two devices is given (similar to our drill example). Here, the previous table height is considered mandatory. Which corresponds to the previous activation state in case of the drill.
This means that the provider has the means to ensure safe concurrent execution by defining an appropriate SafetyReq element. Thus, concurrency can be detected by the provider, **_iff_** necessary. 

MDPWS §9.2 defines the SafetyReq element and provides an example in Annex C.4 how it could be defined. In this example, not the mdib version, but the previous value of the metric is required.
	
So BICEPS R60 requires the provider to guarantee safe concurrent execution.
And MDPWS §9 provides the means for this.

**Thus this issue can be considered solved.**

Note:
This implies, that there is no need to enforce sequential processing! One may use multiple sockets for communication and manufacturers may implement *fully parallelized stacks*. 
Providers have the burden to guarantee safe concurrent execution by requiring appropriate information to be included in the consumer's request.

Note #2:
It may be a good idea to not leave this to the individual manufactuers, but to integrate a mandatory field "mdib version on which the change is applied to" to promote safe use of SDC.

On your question / discussion on how to identify the consumer: Glue R0003 states that the X.509 certificate information should be used to map requests to subscriptions.

**Comment by PaulMartinsen** [comment]
Indeed. Some have found ignoring the tip of the iceberg can have titanic consequences. 

I need to think about this some more, but a couple of initial thoughts:
* using the MDIB version for operation sequencing could be problematic if there are frequent mdib version changes (e.g., 100 ms waveform streams), 
* yes, the 8 minutes I mentioned is the discrepancy that NTPv4 will try to correct with a slewing (smooth) adjustments (that is, by making the local clock run faster or slower). Larger discrepancies would trigger a non-slewing, step adjustment. 
* yes, it is very hard to thinking of a real-world situation where the discrepancy would be this large. The best I came up with so far is a provider with an inaccurate local oscillator that connects to a synchronization server, gets a few reference time-stamps, but not enough to complete the cold start algorithm (which effectively calibrates the local oscillator frequency), loses network connectivity for a long time, reconnects and starts sending biceps messages before it realizes it needs to significantly adjust its clock. Since a 2 minute timeout is prescribed for subscriptions, the local clock oscillator would have to be very bad I think. 
* I can see how including current state helps for settings (e.g., set the drill speed to 4000 RPM, but only if we agree it was 3000  RPM). 
* I haven't been able to see how the drill state problem can be solved through previous state. The failure you raised is where the operator sends drill on -> drill off, but the order gets mixed up and the provider receives drill off -> drill on. Doesn't the drill remain on even with the previous state included when the order is reversed? That is: 
  * drill off (only if it was on) => drill remains off. 
  * drill on (only if it was off) => drill turns on and remains on. 
* As you noted, using the MDIB version as part of the previous state could resolve this... if there weren't a lot of waveform streams in between as well I think.

Thanks for the info on consumer identification; I'd missed that whole section!

**Comment by BenjaminHohlmann** [comment]
You are right on the drill example. In that scenario, just the previous value is not sufficient.
Frequent version changes are indeed an issue: Assuming a latency of 10ms and a version update every 100ms, every 5th set operation will fail:
- the consumer will be "behind" the provider by 10ms due to delay of reports
- the set operation will be delayed by 10ms, too.
This sums up to the set requests being outdated by 20ms on arrival at the provider.

This are very relevant aspect: An update rate of 60Hz does not deem far-fetched, but running such a device on a network with just 9ms latency (network + processing combined!) renders external control (with an mdib-version safety mechanism) impossible!

**Comment by PaulMartinsen** [comment]
Just as I went to sleep I realized there's an obvious solution. If the drill has, for example, an `pm:EnumStringMetricState` reflecting its state (on or off), then perhaps the safety control could be based on the `pm:EnumStringMetricState/@StateVersion` instead of the mdib version. Then it won't be affected by other, unrelated, updates. 

It still presents a challenge for concurrent operation (even from a single consumer) as it forces serialization of operations (parallel operations could have a race condition figuring out what `@StateVersion` to use). And it _may increase risk_ if an operation to stop the drill is blocked by a mismatched `@StateVersion`. Although I suppose there could be separate operations to start (requiring knowledge of `@StateVersion`) and stop (doesn't require `@StateVersion`) that could mitigate this. 

And, of course, network latency is still an issue for turning things on and off quickly if that's needed. I suppose PWM over SDC isn't going to be terribly practical. 

Are you going to participate in the external control PKP meetings Marcus Köny is organizing? This seems relevant for those discussions.

**Comment by BenjaminHohlmann** [comment]
That's indeed a safe and efficient solution!
I will join in.
